### ISSUE TITLE
live multi person editing in godot

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#851]
**Describe the project you are working on:**
Non right now but I would like to work on a project with my older brother and show him godot

**Describe the problem or limitation you are having in your project:**
Me and my brother live in two different countries and I would like to start a rouge lite game with him and introduce him to godot.

the problem is I would like to have live interaction developing this game with him and while git is good it doesn't really acheave the feeling of building something together

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I suggest that the godot editor gets live networked editing similar to what [superpowers](http://superpowers-html5.com/index.en.html) did and [SMBX](https://www.supermariobrosx.org/) tried to do.

this would allow people to work together to build games in a really cool way.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
basically when one user would edit a file or a property the other users would see the changes on the other end.

the way I see this would be something like git being used to save daily work while this would be used to save immediate work.

I would guess that when you connect to someone it does the following..
- lets you choose project if one doesn't have source control it makes you create one
- lets you choose a source control branch. each branch might have different people working on it.
- when something changes for example a property or file it update it in the editor for each person working on that branch (only the changes).
- finally when people are done for the day commits are made to the currently work on branch.  and if everything is implemented right it won't let you commit if your out of sync with others.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
idk if people would use this but it sounds like it could be the next big source control
also it couldn't be worked around.

**Is there a reason why this should be core and not an add-on in the asset library?:**
if implemented right it would speed up development and while it could be implimented as a asset library its very unlikely someone would do something this big

**Comment by RabbitB** [comment]
I genuinely like this idea as a concept, and maybe this is a pessimistic view, but I can only see the amount of work required to make this happen in a fluid and project-safe way, as insurmountable.

**Comment by Shadowblitz16** [comment]
I think the set and get functions would actually make it easier but I still agree it would be alot of work.

**Comment by LikeLakers2** [comment]
It's a great concept, and I've had this idea in my head for a while, but I have to agree with RabbitB here -- it sounds like a little too much work.

However, perhaps you could use something like Dropbox shared folders or a network share to make something similar. You would have to avoid stepping on each other's edits, however, such as ensuring you're both not editing the same scene or the same script file.

Aside from that, perhaps some checks could be added to make something like this a little easier. Perhaps the editor can check, when you save a file, if the file has changed since you opened it/last saved it. And/Or perhaps the editor can create lock files on any files that it opens, so as to indicate to other editors that File_X.gd is open in another process (whether that process lives on the same machine or a different machine).

**Comment by RabbitB** [comment]
If we break it down into the core functionality that needs to be developed, we have:
1. Editor needs network connectivity, and every single change made needs to be broadcast reliably. This is a much larger task than it sounds, as the editor holds a lot of state that we need to track and sync constantly.
2. There needs to be some sort of conflict resolution. In an ideal world, a change to a file would only be made when all clients are fully synced, but the reality is that's not likely to always be the case, and we need conflict resolution on how to resolve this without losing any progress made by anyone. Unlike with git, this is all being updated real-time, so we would need to develop UI to alert developers and provide them the ability to resolve conflicts. I have a suspicion that this could greatly disrupt workflow.
3. The project needs to resync all files every time you launch it. This part isn't that bad, but it still needs to be done.

There's certainly more, but that's a 10,000 ft view of the problem. I don't want to call it impossible because it certainly isn't, but this would be a massive undertaking.

**Comment by Jummit** [comment]
I think this would make a great plugin. There is [this](https://github.com/thimenesup/GodotNetworkedSceneEditor), but I don't know how functional it is.

**Comment by mhilbrunner** [comment]
Yeah, this sounds more like a whole seperate project, scope-wise, as amazing as it would be. Not really feasible, IMO.

**Comment by Shadowblitz16** [comment]
if this was done it would be nice to also see people's mouse cursers

**Comment by Calinou** [comment]
There are a few attempts to implement this feature like [this one](https://github.com/thimenesup/GodotNetworkedSceneEditor), but I think this is too difficult to implement in a safe manner. Not to mention that it may also be quite gimmicky given the effort required to implement it.

For comparison, [Sauerbraten](https://sauerbraten.org) features a cooperative map editor, yet most serious maps are built in singleplayer. This doesn't mean cooperative editing is *never* useful for serious mapping, but it's not used as often as you'd expect.

**Comment by aaronfranke** [comment]
I love the idea of live editing, Google Docs is a blessing for school work and I could see it being really useful for Godot too. However, this could be an add-on in the asset library.

**Comment by Shadowblitz16** [comment]
Note vs code and vs studio now have [live share](https://visualstudio.microsoft.com/services/live-share/) which is the same thing

**Comment by fire** [comment]
I worked out an approach for doing multi-person editing.

Use the Mixer from Ubisoft as the server. The clietns are multiple blender and Godot Engine clients.

https://gitlab.com/ubisoft-animation-studio/mixer/-/issues/8 

We were able to get Mixer's client(protocol, not blender) server portion licensed as MIT compatible.

**Comment by fire** [comment]
https://github.com/ubisoft/mixer/releases/tag/v1.0.0

Ubisoft Mixer is live editing the same scene in Unity and Blender.

The 1.0 release adds skeletal posing and is a major release.

Here are the tutorials 
https://www.youtube.com/channel/UCVfQBSBBvo8GMndYy2TzQHw.

Also Unity compatibility 
https://github.com/ubisoft/vrtist.

This 1.0 release removes the limit of skeletons disappearing for use in Godot Engine.

**Comment by macryc** [comment]
Interesting discussion. What about multiuser collaboration at runtime? Like multiplayer but not quite. Say, you have a mindmap type of app that uses the graphedit node with nodes and connectors. Is there a way to allow more than 1 person move/connect the nodes at the same time and save changes to the same file?

**Comment by Calinou** [comment]
> Is there a way to allow more than 1 person move/connect the nodes at the same time and save changes to the same file?

That's not implemented yet, although it may be feasible with a plugin. This is pretty much what this proposal is about :slightly_smiling_face: 

However, there are a lot of potential issues with desyncs you need to cater for.

**Comment by anunknowperson** [comment]
I described my vision of the solution here: https://github.com/godotengine/godot-proposals/issues/3694

**Comment by TrueSgtMonkey** [comment]
Has this been implemented in Godot 4 at all?

**Comment by Calinou** [comment]
> Has this been implemented in Godot 4 at all?

No.

**Comment by qtqgyt** [comment]
So is this ever planned?

**Comment by starry-abyss** [comment]
I wouldn't wait for this any time soon, because even working alone still has some quirks like editor sometimes nulls all export variables in a scene...

**Comment by Sasino97** [comment]
The only way to do this properly and safely, in my opinion, would be through locking. If I start editing anything in a scene or script, those resources become locked until I save them and also don't perform any action on them for a while (or the scene/script is closed, then it's instantly released). If people working on a multi-user project open a file that is being worked on by someone else, that file needs to show up as locked. The true final version of the project must be separate on a dedicated server, which will also save a backup every time a file is saved, to have a history of versions.

**Comment by anunknowperson** [comment]
@Sasino97 that's just how git works, you don't need to reinvent it.

This PR suggests an implementation similar to the one that was done in Unreal Engine, where multiple users can edit one scene/file over the network like in multiplayer games.

**Comment by Calinou** [comment]
> @Sasino97 that's just how git works, you don't need to reinvent it.

Git does not have a built-in concept of file locking, although Git LFS does to a limited extent. That said, most projects which have high expectations on this aspect will prefer using Perforce instead, and [Godot currently doesn't have a notion of file locking that VCS plugins could make use of](https://github.com/godotengine/godot-proposals/issues/8898).

Regardless, this is technically separate from this proposal. File locking isn't a strict requirement for live multi-user editing (even if you'll probably want it in practice).

**Comment by Shadowblitz16** [comment]
Instead of locking you could do immutable copies.

**Comment by Sasino97** [comment]
> Instead of locking you could do immutable copies.

Wouldn't copies need to be merged eventually? Why would merging changes to the same scene make any sense, for example someone moved a node's position to some place, someone else moved it to somewhere else, what sense would it make to merge those two changes, except if one overwrite the other. In my opinion it makes much more sense to simply have file locking, where you can see in real time the changes being made by someone, but you can't also change the same scene until that someone is done.

**Comment by Sasino97** [comment]
> @Sasino97 that's just how git works, you don't need to reinvent it.

Uhh, that doesn't make much sense. Git works with commits, branches, revisions, merges and so on. Locking is just a small part of it. What I'm suggesting is a full lock until that person is presumably done, without requiring merges. Maybe only a history of revisions.

**Comment by Shadowblitz16** [comment]
That defeats the point of live collaboration.
By definition if you can't edit scripts and the scene at the same time as your friend is it's not live collaboration.

It might work for resources if they were still readable, but scenes and scripts should be treated differently.

**Comment by Aboud04** [comment]
if im not mistaken, this is basically the same thing you’re trying to build/ looking for: https://www.reddit.com/r/godot/comments/1tg7nst/gdcollab_a_live_collaboration_plugin_for_godot/
